File: code-for-my-graduate-paper-cpp-version/src/isdip.cpp
#include "../include/isdip.h"
#include "../include/colorspace_utils.h"
#include <cmath>
#include <algorithm>
#include <set>

ISIDPResult ISDIP::isdip_saliency(double sp_pixels, const cv::Mat& superpixels, 
                                 int sp_num, int width, int height, 
                                 int channels, const cv::Mat& image) {
    
    std::vector<cv::Vec3d> rgb_vals(sp_num);
    std::vector<std::vector<int>> inds(sp_num);
    
    // Calculate mean color for each superpixel
    for (int i = 0; i < sp_num; i++) {
        cv::Vec3d sum(0, 0, 0);
        int count = 0;
        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                if (superpixels.at<int>(y, x) == i + 1) {
                    inds[i].push_back(y * width + x);
                    cv::Vec3d pixel = image.at<cv::Vec3d>(y, x);
                    sum += pixel;
                    count++;
                }
            }
        }
        if (count > 0) rgb_vals[i] = sum / count;
    }
    
    // Color space conversions
    std::vector<cv::Vec3d> lab_vals(sp_num), xyz_vals(sp_num);
    for (int i = 0; i < sp_num; i++) {
        lab_vals[i] = ColorspaceUtils::rgb_to_lab(rgb_vals[i]);
        xyz_vals[i] = ColorspaceUtils::rgb_to_xyz(rgb_vals[i]);
    }
    
    // Construct feature vectors
    std::vector<std::vector<double>> features(sp_num);
    for (int i = 0; i < sp_num; i++) {
        features[i] = {rgb_vals[i][0], rgb_vals[i][1], rgb_vals[i][2],
                      xyz_vals[i][0], xyz_vals[i][1], xyz_vals[i][2],
                      lab_vals[i][0], lab_vals[i][1], lab_vals[i][2]};
    }
    
    int scanning_gap = std::max(1, (int)std::sqrt(sp_pixels) - 1);
    std::vector<std::vector<double>> scanning_result(4, std::vector<double>(sp_num, 0));
    
    // Four direction scanning
    for (int scanning_label = 0; scanning_label < 4; scanning_label++) {
        int loops_count = (scanning_label < 2) ? width : height;
        std::vector<std::vector<int>> bg_seeds(loops_count);
        
        // Generate background seeds for each position
        for (int i = 0; i < loops_count; i += scanning_gap) {
            std::set<int> unique_labels;
            
            if (scanning_label == 0) { // left to right
                for (int y = 0; y < height; y++) {
                    for (int x = 0; x <= i && x < width; x++) {
                        unique_labels.insert(superpixels.at<int>(y, x));
                    }
                }
            } else if (scanning_label == 1) { // right to left
                for (int y = 0; y < height; y++) {
                    for (int x = loops_count - i - 1; x < width; x++) {
                        unique_labels.insert(superpixels.at<int>(y, x));
                    }
                }
            } else if (scanning_label == 2) { // top to bottom
                for (int y = 0; y <= i && y < height; y++) {
                    for (int x = 0; x < width; x++) {
                        unique_labels.insert(superpixels.at<int>(y, x));
                    }
                }
            } else { // bottom to top
                for (int y = loops_count - i - 1; y < height; y++) {
                    for (int x = 0; x < width; x++) {
                        unique_labels.insert(superpixels.at<int>(y, x));
                    }
                }
            }
            
            bg_seeds[i] = std::vector<int>(unique_labels.begin(), unique_labels.end());
        }
        
        std::vector<std::vector<double>> weight_matrix(loops_count, std::vector<double>(sp_num, 0));
        loops_count /= 2;
        
        for (int loop_count = 0; loop_count < loops_count; loop_count += scanning_gap) {
            if (loop_count >= bg_seeds.size()) break;
            
            auto& current_bg_seeds = bg_seeds[loop_count];
            
            // Calculate average background feature
            std::vector<double> sum_bg_feature(9, 0);
            int bg_count = 0;
            for (int label : current_bg_seeds) {
                if (label > 0 && label <= sp_num) {
                    for (int j = 0; j < 9; j++) {
                        sum_bg_feature[j] += features[label - 1][j];
                    }
                    bg_count++;
                }
            }
            
            if (bg_count == 0) continue;
            
            for (int j = 0; j < 9; j++) {
                sum_bg_feature[j] /= bg_count;
            }
            
            // Inherit previous weights
            if (loop_count >= scanning_gap) {
                weight_matrix[loop_count] = weight_matrix[loop_count - scanning_gap];
            }
            
            // Calculate color differences for center seeds
            for (int i = 0; i < sp_num; i++) {
                bool is_bg = std::find(current_bg_seeds.begin(), current_bg_seeds.end(), i + 1) != current_bg_seeds.end();
                if (!is_bg) {
                    double color_diff = calculate_color_distance(features[i], sum_bg_feature);
                    weight_matrix[loop_count][i] += color_diff;
                }
            }
        }
        
        if (loop_count - scanning_gap >= 0 && loop_count - scanning_gap < weight_matrix.size()) {
            scanning_result[scanning_label] = weight_matrix[loop_count - scanning_gap];
        }
    }
    
    return {scanning_result, inds};
}

double ISDIP::calculate_color_distance(const std::vector<double>& f1, const std::vector<double>& f2) {
    double sum = 0;
    for (size_t i = 0; i < f1.size(); i++) {
        double diff = f1[i] - f2[i];
        sum += diff * diff;
    }
    return std::sqrt(sum);
}

File: code-for-my-graduate-paper-python-version/isdip.py
"""
ISDIP (Image Saliency Detection using Iterative Scanning and Directional Propagation)
"""
import numpy as np
from colorspace_utils import rgb_to_lab, rgb_to_xyz

def calculate_color_distance(f1, f2):
    """Calculate Euclidean distance between two feature vectors"""
    diff = np.array(f1) - np.array(f2)
    return np.sqrt(np.sum(diff * diff))

def isdip_saliency(sp_pixels, superpixels, sp_num, width, height, channels, image):
    """
    ISDIP saliency detection algorithm
    Args:
        sp_pixels: average pixels per superpixel
        superpixels: superpixel label matrix (H, W)
        sp_num: number of superpixels
        width, height, channels: image dimensions
        image: input image (H, W, C)
    Returns:
        scanning_result: list of 4 saliency maps from different directions
        inds: list of pixel indices for each superpixel
    """
    # Initialize arrays
    rgb_vals = np.zeros((sp_num, 3))
    inds = [[] for _ in range(sp_num)]
    
    # Calculate mean color for each superpixel
    for i in range(sp_num):
        mask = (superpixels == i + 1)  # 1-based indexing
        y_coords, x_coords = np.where(mask)
        if len(y_coords) > 0:
            # Store pixel indices
            pixel_indices = y_coords * width + x_coords
            inds[i] = pixel_indices.tolist()
            
            # Calculate mean RGB values
            pixel_values = image[y_coords, x_coords]
            rgb_vals[i] = np.mean(pixel_values, axis=0)
    
    # Color space conversions
    lab_vals = np.zeros((sp_num, 3))
    xyz_vals = np.zeros((sp_num, 3))
    
    for i in range(sp_num):
        lab_vals[i] = rgb_to_lab(rgb_vals[i].reshape(1, 1, 3)).flatten()
        xyz_vals[i] = rgb_to_xyz(rgb_vals[i].reshape(1, 1, 3)).flatten()
    
    # Construct feature vectors (RGB + XYZ + LAB)
    features = []
    for i in range(sp_num):
        feature = np.concatenate([
            rgb_vals[i],
            xyz_vals[i], 
            lab_vals[i]
        ])
        features.append(feature)
    

    
    # Scanning parameters
    scanning_gap = max(1, int(np.sqrt(sp_pixels)) - 1)
    scanning_result = [np.zeros(sp_num) for _ in range(4)]
    
    # Four direction scanning
    for scanning_label in range(4):
        if scanning_label < 2:  # horizontal scanning
            loops_count = width
        else:  # vertical scanning
            loops_count = height
            
        # Generate background seeds for each position
        bg_seeds = {}
        for i in range(0, loops_count, scanning_gap):
            unique_labels = set()
            
            if scanning_label == 0:  # left to right
                for y in range(height):
                    for x in range(min(i + 1, width)):
                        unique_labels.add(superpixels[y, x])
            elif scanning_label == 1:  # right to left
                for y in range(height):
                    for x in range(loops_count - i - 1, width):
                        unique_labels.add(superpixels[y, x])
            elif scanning_label == 2:  # top to bottom
                for y in range(min(i + 1, height)):
                    for x in range(width):
                        unique_labels.add(superpixels[y, x])
            else:  # bottom to top
                for y in range(loops_count - i - 1, height):
                    for x in range(width):
                        unique_labels.add(superpixels[y, x])
            
            bg_seeds[i] = list(unique_labels)
        
        # Weight matrix construction
        weight_matrix = {}
        for i in range(0, loops_count, scanning_gap):
            weight_matrix[i] = np.zeros(sp_num)
        
        # Iterate through scanning positions
        loops_count = loops_count // 2
        
        for loop_count in range(0, loops_count, scanning_gap):
            if loop_count not in bg_seeds:
                continue
                
            current_bg_seeds = bg_seeds[loop_count]
            
            # Calculate average background feature
            sum_bg_feature = np.zeros(9)  # RGB + XYZ + LAB = 9 dimensions
            bg_count = 0
            
            for label in current_bg_seeds:
                if label > 0 and label <= sp_num:
                    sum_bg_feature += features[label - 1]  # Convert to 0-based
                    bg_count += 1
            
            if bg_count == 0:
                continue
                
            avg_bg_feature = sum_bg_feature / bg_count
            
            # Inherit previous weights
            if loop_count >= scanning_gap and (loop_count - scanning_gap) in weight_matrix:
                weight_matrix[loop_count] = weight_matrix[loop_count - scanning_gap].copy()
            
            # Calculate color differences for center seeds
            for i in range(sp_num):
                label = i + 1  # Convert to 1-based
                if label not in current_bg_seeds:
                    color_diff = calculate_color_distance(features[i], avg_bg_feature)
                    weight_matrix[loop_count][i] += color_diff
        
        # Store final result
        if loop_count - scanning_gap >= 0 and (loop_count - scanning_gap) in weight_matrix:
            scanning_result[scanning_label] = weight_matrix[loop_count - scanning_gap]
        elif loop_count in weight_matrix:
            scanning_result[scanning_label] = weight_matrix[loop_count]
    
    return scanning_result, inds